File: go_weather/internal/clients/weatherstack.go
package clients

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"goweather/internal/logger"
	"goweather/pkg/types"
)

// WeatherStackClient 
type WeatherStackClient struct {
	APIKey  string
	BaseURL string
	Client  *http.Client
	logger  *logger.Logger
}

func NewWeatherStackClient(apiKey string, timeout time.Duration) *WeatherStackClient {
	return &WeatherStackClient{
		APIKey:  apiKey,
		BaseURL: "http://api.weatherstack.com/current", //HTTP
		Client: &http.Client{
			Timeout: timeout, 
		},
		logger: logger.Get(),
	}
}


func (c *WeatherStackClient) GetWeather(location string) (*types.WeatherStackResponse, error) {
	startTime := time.Now()
	url := fmt.Sprintf("%s?access_key=%s&query=%s", 
		c.BaseURL, c.APIKey, location)

	c.logger.APIRequest("weatherstack", location, url).Msg("API request started")

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		c.logger.APIError("weatherstack", location, err, time.Since(startTime))
		return nil, fmt.Errorf("HTTP isteği oluşturulamadı: %v", err)
	}

	resp, err := c.Client.Do(req)
	responseTime := time.Since(startTime)
	
	if err != nil {
		c.logger.APIError("weatherstack", location, err, responseTime)
		return nil, fmt.Errorf("HTTP isteği başarısız: %v", err)
	}
	defer resp.Body.Close()

	//Status code check
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		apiErr := fmt.Errorf("API hatası (Status: %d): %s", resp.StatusCode, string(body))
		c.logger.APIError("weatherstack", location, apiErr, responseTime)
		return nil, apiErr
	}
	
	c.logger.APIResponse("weatherstack", location, resp.StatusCode, responseTime)

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("Response okunamadı: %v", err)
	}

	var weatherResp types.WeatherStackResponse
	if err := json.Unmarshal(body, &weatherResp); err != nil {
		return nil, fmt.Errorf("JSON parse hatası: %v", err)
	}

	return &weatherResp, nil
}

// Get temperature 
func (c *WeatherStackClient) GetTemperature(location string) (float64, error) {
	weather, err := c.GetWeather(location)
	if err != nil {
		return 0, err
	}

	return weather.Current.Temperature, nil
}

File: node_weather/src/services/weatherstack.js
let got;
const dotenv = require('dotenv');
const { logWeatherRequest } = require('../lib/logger.js');
const { normalizeWeatherStackResponse } = require('../lib/normalize.js');

dotenv.config();
class WeatherStackService {
    constructor() {
        this.apiKey = process.env.WEATHERSTACK_KEY;
        this.baseUrl = process.env.WEATHERSTACK_BASE_URL || 'http://api.weatherstack.com';
        this.timeout = 5000; // Wait 5 s
        
        if (!this.apiKey) throw new Error('WEATHERSTACK_KEY is missing in environment');
    }

    async getWeather(location) {
        const startTime = Date.now();
        
        try {
            if (!got) {
                got = (await import('got')).default;
            }
            const url = `${this.baseUrl}/current`;
            const searchParams = {
                access_key: this.apiKey,
                query: location
            };

            const response = await got(url, {
                searchParams,
                timeout: { request: this.timeout },
                retry: {
                    limit: 2,
                    methods: ['GET']
                }
            }).json();

            const responseTime = Date.now() - startTime;
            
            if (response.error) throw new Error(`WeatherStack error: ${response.error.info}`);
            
            
            const normalizedData = normalizeWeatherStackResponse(response);
            logWeatherRequest('weatherstack', location, responseTime, true);
            
            return normalizedData;
            
        } catch (error) {
            const responseTime = Date.now() - startTime;
            
            logWeatherRequest('weatherstack', location, responseTime, false, error.message);            
            throw new Error(`WeatherStack failed: ${error.message}`);
        }
    }
    async healthCheck() {
        try {
            await this.getWeather('London');
            return true;
        } catch (error) {
            return false;
        }
    }
}

const weatherStackService = new WeatherStackService();
module.exports = { weatherStackService };
